x-pack/plugins/rule_registry/server/alert_data_client/alerts_client.ts

```diff
@@ -0,0 +1,215 @@
+    this.logger = logger;
+    this.authorization = authorization;
+    this.esClient = esClient;
+    this.auditLogger = auditLogger;
+  }
+
+  public async getAlertsIndex(featureIds: string[]) {
+    return this.authorization.getAugmentRuleTypesWithAuthorization(
+      featureIds.length !== 0 ? featureIds : ['apm', 'siem']
+    );
+  }
+
+  private async fetchAlert({ id, index }: GetAlertParams): Promise<AlertType> {
+    try {
+      const result = await this.esClient.search<ParsedTechnicalFields>({
+        // Context: Originally thought of always just searching `.alerts-*` but that could
+        // result in a big performance hit. If the client already knows which index the alert
+        // belongs to, passing in the index will speed things up
+        index: index ?? '.alerts-*',
+        body: { query: { term: { _id: id } } },
+      });
+
+      if (!isValidAlert(result.body.hits.hits[0]._source)) {
+        const errorMessage = `Unable to retrieve alert details for alert with id of "${id}".`;
+        this.logger.debug(errorMessage);
+        throw new Error(errorMessage);
+      }
+
+      return result.body.hits.hits[0]._source;
+    } catch (error) {
+      const errorMessage = `Unable to retrieve alert with id of "${id}".`;
+      this.logger.debug(errorMessage);
+      throw error;
+    }
+  }
+
+  public async get({ id, index }: GetAlertParams): Promise<ParsedTechnicalFields> {
+    try {
+      // first search for the alert by id, then use the alert info to check if user has access to it
+      const alert = await this.fetchAlert({
+        id,
+        index,
+      });
+
+      // this.authorization leverages the alerting plugin's authorization
+      // client exposed to us for reuse
+      await this.authorization.ensureAuthorized({
+        ruleTypeId: alert[RULE_ID],
+        consumer: alert[OWNER],
+        operation: ReadOperations.Get,
+        entity: AlertingAuthorizationEntity.Alert,
+      });
+
+      this.auditLogger?.log(
+        alertAuditEvent({
+          action: AlertAuditAction.GET,
+          id,
+        })
+      );
+
+      return alert;
+    } catch (error) {
+      this.logger.debug(`Error fetching alert with id of "${id}"`);
+      this.auditLogger?.log(
+        alertAuditEvent({
+          action: AlertAuditAction.GET,
+          id,
+          error,
+        })
+      );
+      throw error;
+    }
+  }
+
+  public async update<Params extends AlertTypeParams = never>({
+    id,
+    data,
+    index,
+  }: UpdateOptions<Params>): Promise<ParsedTechnicalFields | null | undefined> {
+    try {
+      const alert = await this.fetchAlert({
+        id,
+        index,
+      });
+
+      await this.authorization.ensureAuthorized({
+        ruleTypeId: alert[RULE_ID],
+        consumer: alert[OWNER],
+        operation: WriteOperations.Update,
+        entity: AlertingAuthorizationEntity.Alert,
+      });
+
+      const updateParameters = {
+        id,
+        index,
+        body: {
```

Comment:
A couple of questions:
1) Do we care about [optimistic concurrency control](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-index_.html#optimistic-concurrency-control-index) for these updates? We are ignoring this now, and it complicates the UX if we start to care about this, but I wanted to make sure the decision was intentional.
2) The update is not waiting for a refresh to happen. This means that the operation will complete faster, but a subsequent search may not show the updated document right away. For reference, the Saved Objects Client uses a refresh setting of `wait_for`. Again, I don't have enough context to know what the right decision is here, but I want to make sure the decision is intentional.

Reply:
We've updated the API to account for optimistic concurrency control as it definitely is a concern. Thanks so much for bringing that up here.